CIP-0139/README.md

```diff
@@ -0,0 +1,77 @@
+---
+CIP: 139
+Title: Plutus Core Builtin Function - `multiIndexArray`
+Category: Plutus
+Status: Proposed
+Authors:
+    - Yura Lazaryev <yuriy.lazaryev@iohk.io>
+    - Philip DiSarro <info@anastasialabs.com>
+Implementors: []
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/1050
+Created: 2025-07-07
+License: CC-BY-4.0
+---
+
+## Abstract
+
+We propose adding a new builtin function, `multiIndexArray`, to Plutus Core. This function takes a list of integer indices and an array, returning a list of the array elements at those positions in the specified order.
+
+## Motivation: why is this CIP necessary?
+
+Plutus Core arrays (CIP-0138) support O(1) individual lookup via `indexArray`. However, extracting multiple elements requires repeated calls to `indexArray`, which:
+
+1. Increases script size and execution cost.
+2. Complicates on-chain logic for batching lookups or reordering.
+3. Prevents efficient bulk access and traversal in a user-defined order.
+
+A single `multiIndexArray` call reduces code and cost overhead by batching lookups and delivering elements in the desired sequence.
+
+## Specification
+
+Add the following builtin function:
+
+```haskell
+multiIndexArray :: forall a. List Integer -> Array a -> List a
+```
+
+- **Inputs**:
+  1. A `List Integer` of zero-based indices, in the order elements should be retrieved.
+  2. An `Array a` to index.
+- **Output**: A `List a` containing the elements at the specified indices, in the same order.
```

Comment:
What happens if the list of indices contains repeated elements?  What order do the elements in the result appear in?  I can see at least four possiblities:

* Prefer the first occurrence of each index: `multiIndexArray [1,2,88,3,2,99, 1] a = [a_1, a_2, a_88, a_99]`
* Prefer the last occurrence of each index:  `multiIndexArray [1,2,88,3,2,99, 1] a = [a_88, a_3, a_2, a_99, a_1]`
* Fail if there's a repeated index
* Return a list containing : `multiIndexArray [1,2,88,3,2,99, 1] a = [a_1, a_2, a_88, a_3, a_2, a_99, a_1]`

The first three of these would all require extra work checking for duplicates, which would slow things down and complicate costing.  The final option seems like the simplest one, but I'm sure it could be objected to as well.